File: src/jzrlib/map/dungeons/DungeonArea.java
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2003 by Ruslan N. Garipov (aka Brigadir).
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih (aka Alchemist).
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package jzrlib.map.dungeons;

import java.util.ArrayList;
import java.util.List;
import jzrlib.core.BaseObject;
import jzrlib.core.Rect;
import jzrlib.map.TileType;
import jzrlib.utils.Logger;

/**
 * 
 * @author Ruslan N. Garipov (aka Brigadir), 2003
 * @author Serg V. Zhdanovskih (aka Alchemist), 2003-2008, 2014-2015
 */
public abstract class DungeonArea extends BaseObject
{
    public final List<DungeonMark> MarksList;
    public final DungeonBuilder Owner;
    public final DungeonMark ParentMark;

    public DungeonArea(DungeonBuilder owner, DungeonMark parentMark)
    {
        this.Owner = owner;
        this.ParentMark = parentMark;
        this.MarksList = new ArrayList<>();
    }

    @Override
    protected void dispose(boolean disposing)
    {
        if (disposing) {
            //this.MarksList.dispose();
        }
        super.dispose(disposing);
    }

    protected final boolean isAllowedMark(DungeonMark mark)
    {
        boolean result;

        try {
            result = !mark.Location.equals(this.ParentMark.Location);

            if (result) {
                int markIterator = 0;
                while (markIterator < this.MarksList.size() & result) {
                    result = !mark.Location.equals((this.MarksList.get(markIterator)).Location);
                    markIterator++;
                }
            }
        } catch (Exception ex) {
            Logger.write("DungeonArea.isAllowedMark(): " + ex.getMessage());
            result = false;
        }

        return result;
    }

    protected final void flushMarksList()
    {
        try {
            for (DungeonMark mark : this.MarksList) {
                int state = mark.getState();
                TileType markType = TileType.ttUndefinedMark;

                switch (state) {
                    case DungeonMark.ms_Undefined:
                        markType = TileType.ttUndefinedMark;
                        break;
                    case DungeonMark.ms_AreaGenerator:
                        markType = TileType.ttAreaGeneratorMark;
                        break;
                    case DungeonMark.ms_RetriesExhaust:
                        markType = TileType.ttRetriesExhaustMark;
                        break;
                    case DungeonMark.ms_PointToOtherArea:
                        markType = TileType.ttPointToOtherAreaMark;
                        break;
                }

                this.Owner.setTile(mark.Location.X, mark.Location.Y, markType);
            }
        } catch (Exception ex) {
            Logger.write("DungeonArea.flushMarksList(): " + ex.getMessage());
            throw ex;
        }
    }

    protected final void tryInsertMark(int aX, int aY, int direction)
    {
        DungeonMark mark = new DungeonMark(this, aX, aY, direction);
        if (this.isAllowedMark(mark)) {
            this.MarksList.add(mark);
        } else {
            mark.dispose();
        }
    }

    public final boolean tryApplyThisArea()
    {
        try {
            boolean result = this.buildArea();
            boolean kill = false;

            this.ParentMark.RetriesLeft = this.ParentMark.RetriesLeft - 1;

            if (result) {
                kill = (this.ParentMark != null && this.ParentMark.isPointsToOtherArea());

                if (kill) {
                    result = false;
                    this.ParentMark.setState(DungeonMark.ms_PointToOtherArea);
                }
            }

            if (!kill) {
                if (result) {
                    result = this.Owner.isFitAreaDimension(this);
                }

                if (!result) {
                    if (this.ParentMark.RetriesLeft > 0) {
                        result = this.tryApplyThisArea();
                    } else {
                        this.ParentMark.setState(DungeonMark.ms_RetriesExhaust);
                    }
                }

                if (result) {
                    this.ParentMark.setState(DungeonMark.ms_AreaGenerator);
                }
            }

            return result;
        } catch (Exception ex) {
            Logger.write("DungeonArea.tryApplyThisArea(): " + ex.getMessage());
            throw ex;
        }
    }

    protected abstract boolean buildArea();

    protected abstract boolean isWallPoint(int ptX, int ptY);

    public abstract boolean isOwnedPoint(int ptX, int ptY);

    public abstract boolean isIntersectWithArea(DungeonArea area);

    public abstract void flushToMap();

    public abstract void generateMarksList();

    public abstract boolean isAllowedPointAsMark(int ptX, int ptY);

    public abstract int getDevourArea();

    public abstract Rect getDimensionRect();
}

File: ZRLib/Map/Dungeons/DungeonArea.cs
/*
 *  "NorseWorld: Ragnarok", a roguelike game for PCs.
 *  Copyright (C) 2003 by Ruslan N. Garipov (aka Brigadir).
 *  Copyright (C) 2002-2008, 2014 by Serg V. Zhdanovskih.
 *
 *  This file is part of "NorseWorld: Ragnarok".
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Collections.Generic;
using BSLib;
using ZRLib.Core;

namespace ZRLib.Map.Dungeons
{
    public abstract class DungeonArea : BaseObject
    {
        public readonly IList<DungeonMark> MarksList;
        public readonly DungeonBuilder Owner;
        public readonly DungeonMark ParentMark;

        protected DungeonArea(DungeonBuilder owner, DungeonMark parentMark)
        {
            Owner = owner;
            ParentMark = parentMark;
            MarksList = new List<DungeonMark>();
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing) {
                //MarksList.dispose();
            }
            base.Dispose(disposing);
        }

        protected bool IsAllowedMark(DungeonMark mark)
        {
            bool result;

            try {
                result = !mark.Location.Equals(ParentMark.Location);

                if (result) {
                    int markIterator = 0;
                    while (markIterator < MarksList.Count & result) {
                        result = !mark.Location.Equals((MarksList[markIterator]).Location);
                        markIterator++;
                    }
                }
            } catch (Exception ex) {
                Logger.Write("DungeonArea.isAllowedMark(): " + ex.Message);
                result = false;
            }

            return result;
        }

        protected void FlushMarksList()
        {
            try {
                foreach (DungeonMark mark in MarksList) {
                    int state = mark.State;
                    TileType markType = TileType.ttUndefinedMark;

                    switch (state) {
                        case DungeonMark.Ms_Undefined:
                            markType = TileType.ttUndefinedMark;
                            break;
                        case DungeonMark.Ms_AreaGenerator:
                            markType = TileType.ttAreaGeneratorMark;
                            break;
                        case DungeonMark.Ms_RetriesExhaust:
                            markType = TileType.ttRetriesExhaustMark;
                            break;
                        case DungeonMark.Ms_PointToOtherArea:
                            markType = TileType.ttPointToOtherAreaMark;
                            break;
                    }

                    Owner.SetTile(mark.Location.X, mark.Location.Y, markType);
                }
            } catch (Exception ex) {
                Logger.Write("DungeonArea.flushMarksList(): " + ex.Message);
                throw ex;
            }
        }

        protected void TryInsertMark(int aX, int aY, int direction)
        {
            DungeonMark mark = new DungeonMark(this, aX, aY, direction);
            if (IsAllowedMark(mark)) {
                MarksList.Add(mark);
            } else {
                mark.Dispose();
            }
        }

        public bool TryApplyThisArea()
        {
            try {
                bool result = BuildArea();
                bool kill = false;

                ParentMark.RetriesLeft = ParentMark.RetriesLeft - 1;

                if (result) {
                    kill = (ParentMark != null && ParentMark.PointsToOtherArea);

                    if (kill) {
                        result = false;
                        ParentMark.State = DungeonMark.Ms_PointToOtherArea;
                    }
                }

                if (!kill) {
                    if (result) {
                        result = Owner.IsFitAreaDimension(this);
                    }

                    if (!result) {
                        if (ParentMark.RetriesLeft > 0) {
                            result = TryApplyThisArea();
                        } else {
                            ParentMark.State = DungeonMark.Ms_RetriesExhaust;
                        }
                    }

                    if (result) {
                        ParentMark.State = DungeonMark.Ms_AreaGenerator;
                    }
                }

                return result;
            } catch (Exception ex) {
                Logger.Write("DungeonArea.tryApplyThisArea(): " + ex.Message);
                throw ex;
            }
        }

        protected abstract bool BuildArea();

        protected abstract bool IsWallPoint(int ptX, int ptY);

        public abstract bool IsOwnedPoint(int ptX, int ptY);

        public abstract bool IsIntersectWithArea(DungeonArea area);

        public abstract void FlushToMap();

        public abstract void GenerateMarksList();

        public abstract bool IsAllowedPointAsMark(int ptX, int ptY);

        public abstract int DevourArea { get; }

        public abstract ExtRect DimensionRect { get; }
    }
}
